### PR TITLE
Limit SharedSecret to 32 byte buffer

### PR DESCRIPTION
**Pull request by tcharding**
Currently `SharedSecret` provides a way to get a shared secret using SHA256 _as well as_ a way to use a custom hash function to get the shared secret. Internally `SharedSecret` uses a 256 byte buffer, this is a tad wasteful. We would like to keep the current functionality but reduce memory usage.

- Patch 1: Pulls the `new_with_hash` logic out into a standalone public function that just returns the 64 bytes representing the x,y co-ordinates of the computed shared secret point. Callers are then responsible for hashing this point to get the shared secret (idea by @Kixunil, thanks).
- Patch 2: Does trivial refactor
- Patch 3: Uses a 32 byte buffer internally for `SharedSecret`. This is basically a revert of the work @elichai did to add the custom hashing logic. @elichai please holla if you are not happy with me walking all over this code :)

### Note to reviewers

Secret obfuscation is done on top of this in https://github.com/rust-bitcoin/rust-secp256k1/pull/396, they could be reviewed in order if this work is of interest to you.

### CODE REVIEW

File: src/ecdh.rs [hunk 1]
```diff
@@ -184,6 +134,12 @@ impl SharedSecret {
     }
 }
 
+impl From<[u8; 32]> for SharedSecret {
+    fn from(inner: [u8; 32]) -> SharedSecret {
+        SharedSecret(inner)
+    }
+}
```
Comment (Kixunil) [inline src/ecdh.rs:141]: It seems like this conversion should be the other way around.

File: src/ecdh.rs [hunk 2]
```diff
@@ -40,75 +40,26 @@ use secp256k1_sys::types::{c_int, c_uchar, c_void};
...
     /// True if the underlying data buffer is empty.
     pub fn is_empty(&self) -> bool {
-        self.data.is_empty()
-    }
-
-    /// Set the length of the object.
-    pub(crate) fn set_len(&mut self, len: usize) {
-        debug_assert!(len <= self.data.len());
-        self.len = len;
+        self.0.is_empty()
```
Comment (Kixunil) [inline src/ecdh.rs:59]: I believe this always returns `false`. Maybe remove the method?

File: src/ecdh.rs [hunk 3]
```diff
@@ -40,75 +40,26 @@ use secp256k1_sys::types::{c_int, c_uchar, c_void};
...
-        &self.data[..self.len]
-    }
-}
-
-impl Deref for SharedSecret {
-    type Target = [u8];
-    fn deref(&self) -> &[u8] {
-        &self.data[..self.len]
-    }
-}
```
Comment (Kixunil) [inline src/ecdh.rs:109]: While `Deref` may be considered anti-pattern, I think `AsRef<[u8]>` should be kept.

**Comment by apoelstra** [comment]
The 256-byte buffer was added in https://github.com/rust-bitcoin/rust-secp256k1/pull/180 which tried to make the shared secret more generic. Unfortunately the discussion there is overwhelmed by safety discussion regarding panics and callbacks. But I think the motivation was that users might implement a callback that provides larger hashes, or serializes the whole point, or something like that.

I agree that 256 seems extraordinarily large and also looks like it might've been a confusion between bits and bytes. I will try to review this PR more carefully today and see if I can understand.

**Comment by Kixunil** [comment]
If custom hashing is still desired then I think `SharedSecret` should be generic with `[u8; 32]` being the default.

> or serializes the whole point

To my knowledge this is a very bad idea and I'm not sure if the library should support it. At least without "dangerous" word being part of the API.

**Comment by elichai** [comment]
The point was to allow this function: https://docs.rs/secp256k1/latest/secp256k1/ecdh/struct.SharedSecret.html#method.new_with_hash to use something other than sha2, it could for example use sha512, or even something bigger(or a bit smaller like RIPEMD160). So I capped it at 256 bytes, there wasn't a specific reason for that number, I just thought that 64 bytes (e.g. sha512) is definitely reasonable, so I gave it a bit more leeway.

As for safety, why is that dangerous? the API requires at least 128 bits, and should be pretty safe AFAIU

**Comment by Kixunil** [comment]
As far as I understand if the point is used directly it could unexpectedly interact with other cryptography and cause vulnerabilities. That's why hashing is required.

**Comment by Kixunil** [comment]
Sure, I have no problem with custom hash functions.

File: src/ecdh.rs (hunk 1 above)
Comment (tcharding) [reply src/ecdh.rs:141]: I don't understand why its the wrong way around? It converts from an array to a shared secret?

Comment (tcharding) [reply src/ecdh.rs:141]: Do you mean you'd like to see `From<SharedSecret> for [u8]`? I added `AsRef` impl as suggested above.

**Comment by tcharding** [comment]
I've implemented all review suggestions but I'm still unsure from the discussion if this PR is a good idea or not?

**Comment by tcharding** [comment]
I messed around a bit with this but cannot work out how to do it in a way that is cleaner than the code is now. If you feel like writing up some ideas please do but its not really that important. I'll just close this after a while if nothing comes of it.

Cheers

**Comment by Kixunil** [comment]
I will try to do it to better understand where the problem is.

**Comment by elichai** [comment]
I'm not sure what's the motivation here or what's the problem you're trying to fix.

Is it not to "waste" 256 bytes for what's usually just 32 bytes?
if so, then:
1. Is that really a problem/bottleneck? 
2. We can try and sprinkle some generics over it or make the custom hash one accept a mutable reference for the output array or something like that.

**Comment by Kixunil** [comment]
Not just bottleneck but also makes code confusing and forces us to track the actual length.

**Comment by apoelstra** [comment]
I wouldn't mind a PR to change the size to 64 bytes, which would let users get the whole point out (I know, a bad idea, but we've made it fairly unergonomic and this is needed for some deployed systems), and then if they actually need some crazy-wide hash they can just pull the explicit point out of the `SharedSecret` and do it themselves.

If anything there is value in reducing "256" because it is a confusing number.

**Comment by Kixunil** [comment]
How about having a standalone function computing the point that doesn't use the `SharedSecret` type? Could be named with `dangerous_` prefix and documentation saying it should be hashed.

**Comment by Kixunil** [comment]
Sorry, I meant to replace the current API with a standalone function. `SharedSecret` would use sha256 only.

**Comment by apoelstra** [comment]
I'd be fine with that.

**Comment by tcharding** [comment]
I've done a total re-write with ideas from the thread of discussion above. The PR message has been re-written. Thanks.

File: src/ecdh.rs [hunk 4]
```diff
@@ -135,52 +135,52 @@ impl SharedSecret {
...
-        let mut x = [0u8; 32];
-        let mut y = [0u8; 32];
-        x.copy_from_slice(&xy[..32]);
-        y.copy_from_slice(&xy[32..]);
-        hash_function(x, y)
-    }
+/// Creates a shared point from public key and secret key.
+///
+/// Can be used like `SharedSecret` but caller is responsible for then hashing the returned buffer.
+/// This allows for the use of a custom hash function since `SharedSecret` uses SHA256.
```
Comment (Kixunil) [inline src/ecdh.rs:86]: I think the wording should be stronger and with emphasis. Something like:

> **Important: use of strong cryptographic hash function may be critical to security! Do NOT use unless you understand cryptographical implications.** If not, use `SharedSecret` instead.

File: src/ecdh.rs [hunk 5]
```diff
@@ -135,52 +135,52 @@ impl SharedSecret {
...
+///
+/// let s = Secp256k1::new();
+/// let (sk1, pk1) = s.generate_keypair(&mut thread_rng());
+/// let (sk2, pk2) = s.generate_keypair(&mut thread_rng());
+///
+/// let point1 = ecdh::shared_secret_point(&pk2, &sk1);
+/// let secret1 = sha512::Hash::hash(&point1);
+/// let point2 = ecdh::shared_secret_point(&pk1, &sk2);
+/// let secret2 = sha512::Hash::hash(&point2);
+/// assert_eq!(secret1, secret2)
```
Comment (Kixunil) [inline src/ecdh.rs:107]: BTW this example could be simplified by comparing the points instead of hashes but maybe there's value in showing hashed usage in the example?

File: src/ecdh.rs [hunk 6]
```diff
@@ -39,90 +38,18 @@ use secp256k1_sys::types::{c_int, c_uchar, c_void};
...
-    }
-}
-
-
-unsafe extern "C" fn c_callback(output: *mut c_uchar, x: *const c_uchar, y: *const c_uchar, _data: *mut c_void) -> c_int {
-    ptr::copy_nonoverlapping(x, output, 32);
-    ptr::copy_nonoverlapping(y, output.offset(32), 32);
-    1
-}
+#[derive(Copy, Clone, Debug, PartialEq)]
```
Comment (Kixunil) [inline src/ecdh.rs:41]: I believe `Debug` should *not* be *derived* but manually censored.

Also, equality is reflexive, so `Eq` could be added as well as `Hash`.

File: src/ecdh.rs [hunk 7]
```diff
@@ -132,11 +59,27 @@ impl SharedSecret {
...
         // and the scalar was verified to be valid(0 > scalar > group_order) via the type system
         debug_assert_eq!(res, 1);
-        ss.set_len(32); // The default hash function is SHA256, which is 32 bytes long.
-        ss
+        SharedSecret(buf)
+    }
+
+    /// Returns true if the underlying data buffer is empty.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
```
Comment (Kixunil) [inline src/ecdh.rs:67]: This will always return `false`, so is it really needed?

File: src/ecdh.rs [hunk 8]
```diff
@@ -132,11 +59,27 @@ impl SharedSecret {
...
     }
 }
 
+impl AsRef<[u8]> for SharedSecret {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
```
Comment (Kixunil) [inline src/ecdh.rs:75]: I think `Borrow` would be also useful.

File: src/ecdh.rs [hunk 9]
```diff
@@ -39,90 +38,18 @@ use secp256k1_sys::types::{c_int, c_uchar, c_void};
...
+pub struct SharedSecret([u8; 32]);
 
 impl SharedSecret {
     /// Creates a new shared secret from a pubkey and secret key.
     #[inline]
     pub fn new(point: &PublicKey, scalar: &SecretKey) -> SharedSecret {
-        let mut ss = SharedSecret::empty();
+        let mut buf = [0u8; 32];
         let res = unsafe {
              ffi::secp256k1_ecdh(
```
Comment (Kixunil) [inline src/ecdh.rs:50]: I'm not sure this is better than calling `shared_secret_point` and then hashing it. If it is, then a test that compares them would be useful.

File: src/ecdh.rs (hunk 1 above)
Comment (Kixunil) [reply src/ecdh.rs:133]: Yes, I mean `From<SharedSecret> for [u8; 32]`

The problem with the opposite conversion is that it's almost implicit. Something like `foo(SharedSecret::from_raw_bytes(bytes))` is clearer than `foo(bytes.into())`.
The newtype is supposed to indicate "this is indeed a shared secret derived using ECDH, not some other 32 bytes coming from god-knows-where"

Comment (tcharding) [reply src/ecdh.rs]: Removed this function all together.

File: src/ecdh.rs (hunk 4 above)
Comment (tcharding) [reply src/ecdh.rs:86]: Cheers, will use.

File: src/ecdh.rs (hunk 5 above)
Comment (tcharding) [reply src/ecdh.rs:107]: Seems to me iff the point _should_ be hashed then the example should show hashing, if it does not matter then we should change the docs to say 'users can hash the point' instead of 'users should hash the point'. I do not know the correct usage?

File: src/ecdh.rs (hunk 6 above)
Comment (tcharding) [reply src/ecdh.rs:41]: Debug is going to be done as a separate PR (#396), will add the other derives.

File: src/ecdh.rs (hunk 7 above)
Comment (tcharding) [reply src/ecdh.rs:67]: Thanks, I incorrectly thought that all zeros would return true. Removed ... and schooled.

File: src/ecdh.rs (hunk 8 above)
Comment (tcharding) [reply src/ecdh.rs:75]: oh yes, this was discussed elsewhere already recently, my bad. Just to clarify, we should use `Borrow` and remove `AsRef`, right?

Comment (tcharding) [reply src/ecdh.rs:75]: We have `src/ecdsa/mod.rs:83:impl AsRef<[u8]> for SerializedSignature {`, should this be changed to implement `Borrow` instead?

File: src/ecdh.rs (hunk 9 above)
Comment (tcharding) [reply src/ecdh.rs:50]: Of course, that is better. Thanks

**Comment by tcharding** [comment]
Thanks @Kixunil, implemented suggestions and force pushed!

File: src/ecdh.rs (hunk 5 above)
Comment (Kixunil) [reply src/ecdh.rs:100]: Hashing is not needed for comparison but yeah, let's keep best practice.

File: src/ecdh.rs (hunk 8 above)
Comment (Kixunil) [reply src/ecdh.rs:71]: No, we should implement both `AsRef` and `Borrow` for shared secret.
For other types we have to evaluate them one-by-one since `Borrow<U> for T` requires that `T::Eq` and `T::Hash` behave exactly same as `U::Eq` and `U::Hash` provided the instance of `U` was obtained by calling `Borrow` on `T`.

File: src/ecdh.rs (hunk 9 above)
Comment (Kixunil) [reply src/ecdh.rs:51]: That's a good reason, I believe. OTOH it annoys me that there are two SHA256 implementations in the code if you enable `bitcoin_hashes`

File: src/ecdh.rs (hunk 8 above)
Comment (tcharding) [reply src/ecdh.rs:71]: > No, we should implement both AsRef and Borrow for shared secret.

Cool, will add.

> since Borrow<U> for T requires that T::Eq and T::Hash behave exactly same as U::Eq and U::Hash provided the instance of U was obtained by calling Borrow on T.

I read that in the docs but I do not understand what it implies. I'll have to take direction from others on this until I do get it.

File: src/ecdh.rs [hunk 10]
```diff
@@ -207,38 +149,6 @@ mod tests {
...
-            x_out = x;
-            y_out = y;
-            expect_result.into()
-        });
-        assert_eq!(&expect_result[..], &result[..]);
-        assert_ne!(x_out, [0u8; 32]);
-        assert_ne!(y_out, [0u8; 32]);
-    }
-
     #[test]
```
[the remark this answers is not in the document]
Comment (tcharding) [reply src/ecdh.rs:152]: Not sure what this passing test says about the implementation in this PR?
```
    #[test]
    #[cfg(all(feature="rand-std", feature = "std", feature = "bitcoin_hashes"))]
    fn bitcoin_hashes_and_sys_generate_same_secret() {
        use hashes::{sha256, Hash, HashEngine};

        let s = Secp256k1::signing_only();
        let (sk1, _) = s.generate_keypair(&mut thread_rng());
        let (_, pk2) = s.generate_keypair(&mut thread_rng());

        let secret_sys = SharedSecret::new(&pk2, &sk1);

        let point = shared_secret_point(&pk2, &sk1);

        let version = (point[31] & 0x01) | 0x02;
        let mut engine = sha256::HashEngine::default();
        engine.input(&[version]);
        engine.input(&point.as_ref()[..32]);
        let secret_bh = sha256::Hash::from_engine(engine);

        assert_eq!(secret_bh.as_inner(), secret_sys.as_ref());
    }
```

File: src/ecdh.rs [hunk 11]
```diff
@@ -132,55 +60,69 @@ impl SharedSecret {
...
+            ffi::secp256k1_context_no_precomp,
+            xy.as_mut_ptr(),
+            point.as_ptr(),
+            scalar.as_ptr(),
+            Some(c_callback),
+            ptr::null_mut(),
+        )
+    };
+    // Our callback *always* returns 1.
+    // The scalar was verified to be valid (0 > scalar > group_order) via the type system.
```
[the remark this answers is not in the document]
Comment (tcharding) [reply src/ecdh.rs:123]: I'm removing this comment totally, I don't see how the type system (type is just a wrapper around 32 bytes) enforces a value less than the group order. My bad for blindly cut'n'pasting before. Bad Tobin, no biscuit.

File: src/ecdh.rs (hunk 9 above)
Comment (apoelstra) [reply src/ecdh.rs:51]: Do you mean, the implementation inside libsecp as well as the implementation in bitcoin_hashes? That is a problem for all users of libsecp, even upstream, and has been the subject of much discussion e.g. https://github.com/bitcoin-core/secp256k1/issues/702

For now there's nothing we can do about it from the Rust bindings. (Well, ok, we could actually expose the hash functionality ourselves by patching the vendored library, but I think the solution we'd rather have is to somehow use the `bitcoin_hashes` implementations when available.)

Comment (Kixunil) [reply src/ecdh.rs:51]: Yes, I meant it. Would be a nice change later if/when it's available.

File: src/ecdh.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (tcharding) [reply src/ecdh.rs:156]: I'm not sure what this test says about the current implementation?:
```
    #[test]
    #[cfg(all(feature="rand-std", feature = "std", feature = "bitcoin_hashes"))]
    fn bitcoin_hashes_and_sys_generate_same_secret() {
        use hashes::{sha256, Hash, HashEngine};

        let s = Secp256k1::signing_only();
        let (sk1, _) = s.generate_keypair(&mut thread_rng());
        let (_, pk2) = s.generate_keypair(&mut thread_rng());

        let secret_sys = SharedSecret::new(&pk2, &sk1);

        let point = shared_secret_point(&pk2, &sk1);

        let version = (point[63] & 0x01) | 0x02;
        let mut engine = sha256::HashEngine::default();
        engine.input(&[version]);
        engine.input(&point.as_ref()[..32]);
        let secret_bh = sha256::Hash::from_engine(engine);

        assert_eq!(secret_bh.as_inner(), secret_sys.as_ref());
    }
```

Also, this passes locally but on CI something is up, `secret_bh` has the value: [85, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]?

[the remark this answers is not in the document]
Comment (Kixunil) [reply src/ecdh.rs:156]: If it behaves differently then something is likely wrong with the code. Not sure why and can't promise having time to debug it.

[the remark this answers is not in the document]
Comment (tcharding) [reply src/ecdh.rs:156]: No worries, I posted just in case the 85,0,0,0 ... meant something immediately to anyone.

File: src/ecdh.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Kixunil) [reply src/ecdh.rs:121]: As far as I can see `SecretKey` **is** verified to be a valid scalar. If you see a constructor that doesn't verify it (didn't check all, just two) it is most likely a bug.

File: src/ecdh.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (tcharding) [reply src/ecdh.rs:156]: It fails during fuzzing, I have not groked the fuzzing stuff yet but this comment in `key.rs` gives a clue. If its quick to answer can someone explain this please.

> In fuzzing mode secret->public key derivation is different

File: no_std_test/src/main.rs [hunk 12]
```diff
@@ -125,13 +125,7 @@ fn start(_argc: isize, _argv: *const *const u8) -> isize {
     assert_eq!(sig, new_sig);
 
     let _ = SharedSecret::new(&public_key, &secret_key);
-    let mut x_arr = [0u8; 32];
-    let y_arr = SharedSecret::new_with_hash(&public_key, &secret_key, |x,y| {
-        x_arr = x;
-        y.into()
-    });
-    assert_ne!(x_arr, [0u8; 32]);
-    assert_ne!(&y_arr[..], &[0u8; 32][..]);
```
Comment (apoelstra) [inline no_std_test/src/main.rs:134]: In 834f63c26cf8b479c1fb288f9887ab72221b3303:

I think you coulrd've kept these asserts in place.

**Review by apoelstra — APPROVED** [review]
ACK 5603d71ad3057ab2572c508b057b7e10f9ae0595

File: src/ecdh.rs (hunk 8 above)
Comment (Kixunil) [reply src/ecdh.rs:69]: Your code is correct. It would've been incorrect if `SharedSecret` overrode `Hash` or `Eq` implementation to behave differently than one on the slice it returns.

File: no_std_test/src/main.rs (hunk 12 above)
Comment (tcharding) [reply no_std_test/src/main.rs:134]: When removing these I could not see what functionality they were testing that was specific to `no-std`, so I removed them. Or am I missing something? If a test that checks the point returned is non-zero is useful I can add it as a unit test, but again, I don't understand the value of it.

Comment (apoelstra) [reply no_std_test/src/main.rs:134]: I think they were just there as a sanity check. I agree they don't seem to have a very clear value (and also that they should be a unit test, if they exist at all).

**Comment by apoelstra** [comment]
@Kixunil do you have open issues here or can I merge this?

**Comment by Kixunil** [comment]
I don't see any obvious defect but wasn't able to review *deeply*. The change is quite simple so I guess that's OK.

**Comment by dspicher** [comment]
This PR [removed](https://github.com/rust-bitcoin/rust-secp256k1/pull/402/files#diff-2ada98d540e88906bfe287caa999ddd6afa38d46fc8733ce014857d4cdd5801eL52) the possibility to create a `SharedSecret` from a byte serialization, also from the still applying `[u8; 32]`. Was this intentional?

**Comment by Kixunil** [comment]
@dspicher good question. I don't think it was intentional but I don't see a huge harm. Why would anyone pretend that some arbitrary bytes are shared secret? Why would anyone compute the actual secret differently than calling `SharedSecret::new()`?

**Comment by Kixunil** [comment]
I'm not sure if it's a good idea to store `SharedSecret` but that use case sounds reasonable. You could also just turn `SharedSecret` into an array or better your own specific newtype. Still seems better to support it.

**Comment by apoelstra** [comment]
Yeah, oops, we should support this. FWIW I think you can get this effect through the serde API.

**Comment by apoelstra** [comment]
Oh, it looks like we don't even have serde de/serialization for this type. We should.
